feat(common):  Support of optional keys for the KeyValue pipe

JeanMeche opened the pull request:
This commit is extending the capabilities of the KeyValue pipe by allowing interfaces with optional keys.

fixes #46867

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [x] No

walidBouguima commented on packages/common/test/pipes/keyvalue_pipe_spec.ts:
```diff
@@ -84,6 +84,28 @@ describe('KeyValuePipe', () => {
     });
+
+    it('should accept an object with optional keys', () => {
+      interface MyInterface {
+        one?: string;
+        two?: number;
+        three?: string;
+        four?: string;
+      }
```
Let's probably use TypeScript's utility Types  (`Partial<Interface>`) instead of manually set every property of the interface to optional with **'?'** .
```suggestion
      interface MyInterface {
        one: string;
        two: number;
        three: string;
        four: string;
      }
```

walidBouguima commented on packages/common/test/pipes/keyvalue_pipe_spec.ts:
```diff
@@ -84,6 +84,28 @@ describe('KeyValuePipe', () => {
+        three?: string;
+        four?: string;
+      }
+      const myData: MyInterface = {
```
```suggestion
      const myData: Partial<MyInterface> = {
```

dylhunn commented on packages/common/src/pipes/keyvalue_pipe.ts:
```diff
@@ -82,6 +82,8 @@ export class KeyValuePipe implements PipeTransform {
       input: Record<K, V>|ReadonlyMap<K, V>|null|undefined,
       compareFn?: (a: KeyValue<K, V>, b: KeyValue<K, V>) => number): Array<KeyValue<K, V>>|null;
+  transform<T, K extends keyof T>(input: T, compareFn?: (a: T[K], b: T[K]) => number):
+      T extends null? null: Array<T[K]>;
```
Why is the conditional type `T extends null? null: Array<T[K]>` necessary as the return type? e.g. why not just `Array<T[K]>`?

JeanMeche replied:
It's in case of a union with null. 
```
declare const myData: MyInterface | null;
new KeyValuePipe().transform(myData); // never[] without the conditional return type
```

[Playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgaWAE8A1TAGwFdgAefAGjhID44BvAKDm7gGtiAXAQDcXHgDcK1ISVEBfDh1CRYcbOUwBnTQWJkqwAAoIweTjzUQkmmFErYY0ABQBKdnAWLutzNfTQAWzpGFicxC2QwShghACVgTAATK3IiAFlMMGCmZnpwnmwIALBMKGAAMSQAfiEnTCFCUilaBhzGACMGvWbsljcAXlYkSgD2tBchAEEoKEwiOm6DXuZmUW9Zv0C6OFBUJESdYdG0EOYwi25I6LjgQqhE5cZC4tKK6tr63SalmyhkAHNTh0ut9qDRfgDTgMhiMxlAJvluNNZvNGvowRCkICcqsxD5NlAgvgdiA9gc4JjsaFEYgkFEYnB4ncHq0WAAfeJJFLpTKPGnPEplSo1OB1EHolpAuCdL4S5bQuBHOETODIuYLUGSnFrOD4zT+QnnC5XBnDcjkNmUfbAdDIYCJPIXSwvIXvUWfNE9K28JAQADuSEY3t9Ady0vFXqQPv9gbgwZjzAVSvGQjN5B1eoNRNORoidOujISySQqQyWVZzDZact1ttSHtjouAtewo+EaWFeBsp6FaTsJTqpm6s9HdOlbTGY2+q2xN2wH2h37UBzNJNN2ZyyrlHNNcSNrtDv5RUFbxFYu7P1skLa4YvGKvWKhcEGiqXCKdatRi3vf0fOK35qTr406Etsc4LhSD5Umcq75gyTLQCypwckW3Jlpu1ZWnudYNkeLqnm2d5aiwXYjmCvbPjCxzwlMQ5fpqyzjtu6Z4lOWY0AAKows6kvO5L8EQEDoHAHFnGuIlPMeLZuueHEANr4AAul28lKX21Eqp+nEKYpqxwAA9PpcAAOoIDAAAWEDRLq5l4IU+xmQgVgUHAZQwJQUBILqRCmKxwHsRWuY8OJWH7vWiQARabByQJQiUopMiMDFgivtRCVMHIKFciWPLlkCeEnpUhFkcRYYyiV8qUalyoCDSWkVRWTHmuwNLcG5HleXJilwFoPVIEQ+QKJ4yCoBgOB4GkRAAJIoGgWC4C1FhWMAIqUjq3h+hAq1QetNllCtcU7Rwnh7uoryWNY8ABEQAAimAwJ8k0zaN814Gyr6ARw9k2HAYAmHg-SKsAfpEcYpiuKI33wP4EDPr9-0AHSZoETjXXdD0uBwhkWAAelUQA)

JeanMeche commented:
Any interest in reviewing this PR ? It can be usefull and doesn't add anything to the bundle 😄

jessicajaniuk reviewed (approved):
LGTM!

reviewed-for: fw-common, public-api

jessicajaniuk commented:
@JeanMeche There's a little lack of clarity on the types union here and how it solves the problem. Can you add some documentation on it to clarify how it solves the problem so it's easier for reviewers?

thePunderWoman reviewed (approved):
reviewed-for: public-api
LGTM

JeanMeche committed:
feat(common): Support of optional keys for the KeyValue pipe

This commit is extending the capabilities of the KeyValue pipe by allowing interfaces with optional keys.

fixes angular#46867
